smoke-tester/src/main.rs

```diff
@@ -123,88 +128,102 @@ fn main() -> Result<ExitCode> {
-        //       some cores need additional setup so that they can be used, and this is not handled yet.
-        for (core_index, core_type) in cores.into_iter().take(1) {
-            println_dut_status!(tracker, blue, "Core {}: {:?}", core_index, core_type);
+            macro_rules! run_test {
+                (generic, $test_fn:ident, $tracker:ident, $($param:tt),* ) => {
+                    let result = $tracker
+                        .run_test(async |tracker| $test_fn(tracker, $($param),*).await)
+                        .await;
 
-            let mut core = session.core(core_index)?;
+                    if let Err(TestFailure::Fatal(error)) = result.result {
+                        return Err(error.context("Fatal error in test"));
+                    }
 
-            println_dut_status!(tracker, blue, "Halting core..");
+                    if result.failed() {
+                        fail_counter += 1;
+                    }
+                };
+                (core, $test_fn:ident, $tracker:ident, $core:ident ) => {
+                    let c = &mut $core;
+                    run_test!(generic, $test_fn, $tracker, c)
+                };
+                (session, $test_fn:ident, $tracker:ident, $session:ident ) => {
+                    let s = &mut $session;
+                    run_test!(generic, $test_fn, $tracker, s)
+                };
 
-            core.reset_and_halt(Duration::from_millis(500))?;
+            }
 
-            for test_fn in CORE_TESTS {
-                let result = tracker.run_test(|tracker| test_fn(tracker, &mut core));
+            // TODO: Handle different cores. Handling multiple cores is not supported properly yet,
+            //       some cores need additional setup so that they can be used, and this is not handled yet.
+            for (core_index, core_type) in cores.into_iter().take(1) {
+                println_dut_status!(tracker, blue, "Core {}: {:?}", core_index, core_type);
 
-                if let Err(TestFailure::Fatal(error)) = result.result {
-                    return Err(error.context("Fatal error in test"));
-                }
+                let mut core = session.core(core_index)?;
 
-                if result.failed() {
-                    fail_counter += 1;
-                }
-            }
+                println_dut_status!(tracker, blue, "Halting core..");
 
-            // Ensure core is not running anymore.
-            core.reset_and_halt(Duration::from_millis(200))
-                .with_context(|| {
-                    format!("Failed to reset core with index {core_index} after test")
-                })?;
-        }
+                core.reset_and_halt(Duration::from_millis(500))?;
 
-        for test in SESSION_TESTS {
-            let result = tracker.run_test(|tracker| test(tracker, &mut session));
+                run_test!(core, test_register_read, tracker, core);
```

Comment:
We have to explicitely run the tests currently because `distributed_slice` of `linkme` does not work with AsyncFn because it as a generic that is a compiler internal that we cannot name currently.